src/main/java/com/exactpro/th2/mstore/AbstractMessageStore.java

```diff
@@ -59,35 +73,105 @@ public final void handle(T delivery) {
             AtomicLong lastSeqHolder = sessionToLastSequence.computeIfAbsent(sessionKey, ignore -> new AtomicLong(Long.MIN_VALUE));
             long prevLastSeq = lastSeqHolder.getAndAccumulate(lastSequence, Math::max);
             if (prevLastSeq >= firstSequence) {
-                logger.error("Duplicated delivery found: {}", delivery);
+                logger.error("Duplicated delivery found: {}", messageBatch);
                 return;
             }
             storeMessages(messages);
         } catch (Exception ex) {
-            logger.error("Cannot handle delivery of type {}", delivery.getClass(), ex);
+            logger.error("Cannot handle delivery of type {}", messageBatch.getClass(), ex);
         }
     }
 
-    protected void storeMessages(List<M> messagesList) throws CradleStorageException, IOException {
+    @Override
+    public void dispose() {
+        super.dispose();
+
+        logger.debug("Waiting for futures completion");
+        Collection<CompletableFuture<Void>> futuresToRemove = new HashSet<>();
+        while (!(asyncStoreFutures.isEmpty() || Thread.currentThread().isInterrupted())) {
+            logger.info("Wait for the completion of {} futures", asyncStoreFutures.size());
+            futuresToRemove.clear();
+            asyncStoreFutures.forEach((future, batch) -> {
+                try {
+                    if (!future.isCancelled()) {
+                        future.get(1, TimeUnit.SECONDS);
+                    }
+                    futuresToRemove.add(future);
+                } catch (InterruptedException e) {
+                    logger.error(e.getMessage(), e);
+                    Thread.currentThread().interrupt();
+                } catch (CancellationException e) {
+                    if (logger.isWarnEnabled()) {
+                        logger.warn("{} - future related to {} batch is cancelled", getClass().getSimpleName(), formatStoredMessageBatch(batch, false), e);
+                    }
+                } catch (ExecutionException e) {
+                    if (logger.isWarnEnabled()) {
+                        logger.warn("{} - storing {} batch is failure", getClass().getSimpleName(), formatStoredMessageBatch(batch, false), e);
+                    }
+                } catch (TimeoutException e) {
+                    if (logger.isErrorEnabled()) {
+                        logger.error("{} - future related to {} batch can't be complited", getClass().getSimpleName(), formatStoredMessageBatch(batch, false), e);
+                    }
+                    future.cancel(false);
+                }
```

Comment:
I believe that there is some executor inside the cradle that executes all our requests. Probably, it should provide the ability to shutdown it and wait for all already submitted tasks are completed